lightningd/peer_control.c

```diff
@@ -1343,6 +1343,89 @@ static const struct json_command getinfo_command = {
 };
 AUTODATA(json_command, &getinfo_command);
 
+
+static struct command_result *json_setchannelfee(struct command *cmd,
+					 const char *buffer,
+					 const jsmntok_t *obj UNNEEDED,
+					 const jsmntok_t *params)
+{
+	struct json_stream *response;
+	const jsmntok_t *idtok;
+	unsigned int *base;
+	unsigned int *ppm;
+	struct peer *peer;
+	/* FIXME: gcc 7.3.0 thinks this might not be initialized. */
+	struct channel *channel = NULL;
+	struct channel_id cid;
+	struct command_result *res;
+
+	if (!param(cmd, buffer, params,
+		   p_req("id", param_tok, &idtok),
+		   p_opt_def("base", param_number, &base, 0),
+		   p_opt_def("ppm", param_number, &ppm, 0),
+		   NULL))
+		return command_param_failed();
+
+	/* find the channel */
+	peer = peer_from_json(cmd->ld, buffer, idtok);
+	if (peer) {
+		channel = peer_active_channel(peer);
+		if (!channel)
+			return command_fail(cmd, LIGHTNINGD,
+				    "Could not find active channel of peer with that id");
+	} else {
+		res = command_find_channel(cmd, buffer, idtok, &channel);
+		if (res)
+			return res;
+	}
+
+	/* check channel is found and in valid state */
+	if (!channel)
+		return command_fail(cmd, LIGHTNINGD, "Could not find peer with that id");
+	if (channel->state != CHANNELD_NORMAL)
+		return command_fail(cmd, LIGHTNINGD, "Channel is not in state CHANNELD_NORMAL");
+
+	/* set, notify and save new values */
+	channel->feerate_base = *base;
+	channel->feerate_ppm = *ppm;
+
+	/* tell channeld to make a send_channel_update */
+	if (channel->owner && streq(channel->owner->name, "lightning_channeld"))
+		subd_send_msg(channel->owner,
+				  take(towire_channel_json_feerates(channel, *base, *ppm)));
+
+	/* save values to database */
+	wallet_channel_save(cmd->ld->wallet, channel);
+
+	/* get channel id */
+	derive_channel_id(&cid, &channel->funding_txid, channel->funding_outnum);
+
+	/* write response */
+	response = json_stream_success(cmd);
+	json_object_start(response, NULL);
+	json_add_string(response, "channel_id",
+		type_to_string(tmpctx, struct channel_id, &cid));
+	if (channel->scid)
+		json_add_short_channel_id(response, "short_channel_id", channel->scid);
+	if (*base > 0)
+		json_add_num(response, "base", *base);
+	if (*ppm > 0)
+		json_add_num(response, "ppm", *ppm);
+	json_object_end(response);
+	return command_success(cmd, response);
+}
+
+static const struct json_command setchannelfee_command = {
+	"setchannelfee",
+	json_setchannelfee,
+	"Sets specific routing fees for channel with {id} "
+	"(either peer ID, or short channel ID). "
+	"The routing fees are calclulated by a fixed {base} "
+	"and a proportional {ppm}. Params are in MSAT, "
```

Comment:
parameters are not in msat (well, base is).  You *could* use param_msat for that, in fact (which lets them specify it in sat if they want, for example), but if you do make sure you check it doesn't exceed 32 bits!

Reply:
yes, thx. I now do the 32bit overflow check and use `base->millisatoshis`, works but complains about code style: `[Makefile:332: check-amount-access] Error 1`

How am I supposed to access a checked value?

Reply:
update: probably not like this :D
```C
u32 value = *(u32*)base;
```